### PR TITLE
feat: Add set operation (select ... union select ...) for insert stat…

### PR DESCRIPTION
**Pull request by antoineB**
…ements.

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -489,7 +496,7 @@ module.exports = grammar({
         $.keyword_regproc,
         $.keyword_regtype,
 
-        field("custom_type", $._identifier)
+        field("custom_type", $.object_reference)
```
Comment (dmfay) [inline grammar.js:499]: are there any databases in which types belong to schemata? if not `_identifier` is correct but could be aliased

File: grammar.js [hunk 2]
```diff
@@ -551,8 +558,21 @@ module.exports = grammar({
     nchar: $ => parametric_type($, $.keyword_nchar),
     nvarchar: $ => parametric_type($, $.keyword_nvarchar),
 
+    _time_zone: $ => seq(
```
Comment (dmfay) [inline grammar.js:561]:
```suggestion
    _include_time_zone: $ => seq(
```
this is clearer imo, but I definitely like the more detailed deconstruction of timestamps overall!

**Review by dmfay — COMMENTED** [review]
something interesting's happening here: state count is much higher in this branch (7776 vs 4536 on main), which I think is largely down to the addition of function argument modes -- `invocation` punches way above its weight class (big increases also in `binary_expression`, `between_expression`, `implicit_cast`, and a bunch of types like `double` and `tinyint` -- maybe due to the time additions?). Large state count has _dropped_ by a lot, though, 775 in main to 333 here, and build time from an empty src/ is actually around ten seconds faster on my machine. Do you know what's happening here?

also if you have more changes in the mean time please create separate pull requests! this one's getting unwieldy

File: grammar.js (hunk 1 above)
Comment (antoineB) [reply grammar.js:499]: It is the case in postgres https://www.postgresql.org/docs/current/sql-createtype.html.

File: grammar.js (hunk 2 above)
Comment (antoineB) [reply grammar.js:561]: Fine by me.

**Comment by matthias-Q** [comment]
The state count is shown in the generated parser.c

**Comment by dmfay** [comment]
yeah, `head -18 src/parser.c` shows you all the potentially useful counts.

File: grammar.js (hunk 1 above)
Comment (dmfay) [reply grammar.js:499]: interesting, I didn't know that!

**Comment by antoineB** [comment]
Here is all the numbers

f75b8ea
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7637
#define LARGE_STATE_COUNT 328
#define SYMBOL_COUNT 702
#define ALIAS_COUNT 0
#define TOKEN_COUNT 407
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 197

531d243
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7528
#define LARGE_STATE_COUNT 323
#define SYMBOL_COUNT 702
#define ALIAS_COUNT 0
#define TOKEN_COUNT 407
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

bec206c
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7362
#define LARGE_STATE_COUNT 317
#define SYMBOL_COUNT 698
#define ALIAS_COUNT 0
#define TOKEN_COUNT 406
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

c223823
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7362
#define LARGE_STATE_COUNT 317
#define SYMBOL_COUNT 698
#define ALIAS_COUNT 0
#define TOKEN_COUNT 406
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

1e3f464
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7362
#define LARGE_STATE_COUNT 317
#define SYMBOL_COUNT 698
#define ALIAS_COUNT 0
#define TOKEN_COUNT 406
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

bde914f
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7362
#define LARGE_STATE_COUNT 317
#define SYMBOL_COUNT 698
#define ALIAS_COUNT 0
#define TOKEN_COUNT 406
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

1a2b0da
1,4M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 7373
#define LARGE_STATE_COUNT 322
#define SYMBOL_COUNT 696
#define ALIAS_COUNT 0
#define TOKEN_COUNT 406
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

970b548
1,6M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 4590
#define LARGE_STATE_COUNT 787
#define SYMBOL_COUNT 670
#define ALIAS_COUNT 0
#define TOKEN_COUNT 386
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

385aff4 (main)
1,5M	target/parser.so
#define LANGUAGE_VERSION 14
#define STATE_COUNT 4536
#define LARGE_STATE_COUNT 775
#define SYMBOL_COUNT 670
#define ALIAS_COUNT 0
#define TOKEN_COUNT 386
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 42
#define MAX_ALIAS_SEQUENCE_LENGTH 16
#define PRODUCTION_ID_COUNT 190

File: grammar.js [hunk 3]
```diff
@@ -262,6 +265,28 @@ module.exports = grammar({
...
     keyword_support: _ => make_keyword("support"),
+    keyword_extension: _ => make_keyword("extension"),
+    keyword_out: _ => make_keyword("out"),
+    keyword_inout: _ => make_keyword("inout"),
+    keyword_variadic: _ => make_keyword("variadic"),
+
+    keyword_session: _ => make_keyword("session"),
+    keyword_isolation: _ => make_keyword("isolation"),
+    keyword_level: _ => make_keyword("level"),
+    keyword_serializable: _ => make_keyword("seria"),
```
Comment (matthias-Q) [inline grammar.js:276]: is this called seria? I have search pg and mariadb docs and was not able to find this. Can you give me an example?

File: grammar.js [hunk 4]
```diff
@@ -262,6 +265,28 @@ module.exports = grammar({
...
+    keyword_repeatable: _ => make_keyword("repeatable: _ ="),
+    keyword_read: _ => make_keyword("read: _ =>"),
```
Comment (matthias-Q) [inline grammar.js:278]: Again, I was not able to fin this structure in PG and MariaDB.  Also separating tokens in the keyword statement is not the right way. In SQL there can be multiple spaces/tabs between tokens.

File: grammar.js [hunk 5]
```diff
@@ -262,6 +265,28 @@ module.exports = grammar({
...
+    keyword_committed: _ => make_keyword("committed"),
+    keyword_uncommitted: _ => make_keyword("uncommitted"),
+    keyword_deferrable: _ => make_keyword("deferrable"),
+    keyword_names: _ => make_keyword("names"),
+    keyword_zone: _ => make_keyword("zone"),
+    keyword_immediate: _ => make_keyword("immediate"),
+    keyword_deferred: _ => make_keyword("deferred"),
+    keyword_constraints    : _ => make_keyword("constraints"),
+    keyword_snapshot: _ => make_keyword("snapshot"),
+    keyword_characteristics: _ => make_keyword("characteristics"),
```
Comment (matthias-Q) [inline grammar.js:289]: Most of these Keywords are not used at the moment, correct? I would prefer only adding those that are actually used in the parser.

File: grammar.js [hunk 6]
```diff
@@ -735,6 +757,31 @@ module.exports = grammar({
       ),
     ),
 
+    _argmode: $ => choice(
```
Comment (matthias-Q) [inline grammar.js:760]: can you add one or two test cases for this?

File: grammar.js [hunk 7]
```diff
@@ -904,6 +961,54 @@ module.exports = grammar({
       ),
     ),
 
+    reset_statement: $ => seq(
```
Comment (matthias-Q) [inline grammar.js:964]: Again, can you add test cases here?

File: grammar.js [hunk 8]
```diff
@@ -904,6 +961,54 @@ module.exports = grammar({
...
+      ),
+      choice(
+        seq($.keyword_read, $.keyword_write),
+        seq($.keyword_read, $.keyword_only),
+      ),
+      optional($.keyword_not),
+      $.keyword_deferrable,
+    ),
+
+    set_statement: $ => seq(
```
Comment (matthias-Q) [inline grammar.js:991]: Same as above. Test cases please.

File: test/corpus/insert.txt [hunk 9]
```diff
@@ -526,3 +526,38 @@ FROM
...
+
+================================================================================
+Insert from unioned select
+================================================================================
+
+INSERT INTO some_table
+  (field)
+  (SELECT "String value"
+   UNION
+   SELECT "String value");
```
Comment (matthias-Q) [inline test/corpus/insert.txt:538]: For pg these should be single quotes, right?

**Review by matthias-Q — CHANGES_REQUESTED** [review]
Hi, thanks for you PR. I have made some remarks. Would be nice If you could address them.

For future PRs, I would be nice if you could limit the scope of the PR to one topic and don't let the PR get too large. I makes it hard to review. If I am not mistaken, most of the suggested changes have nothing to do with the title of the PR either. A (very short) description of what you want to address/change/implement is also welcome.

File: grammar.js (hunk 3 above)
Comment (antoineB) [reply grammar.js:276]: It is a typo.

File: grammar.js (hunk 4 above)
Comment (antoineB) [reply grammar.js:278]: Same as before a keyboard macro typo.

File: grammar.js (hunk 5 above)
Comment (antoineB) [reply grammar.js:289]: They are used in set_statement

File: test/corpus/insert.txt (hunk 9 above)
Comment (antoineB) [reply test/corpus/insert.txt:538]: Yes

File: grammar.js (hunk 6 above)
Comment (antoineB) [reply grammar.js:761]: done

File: grammar.js (hunk 7 above)
Comment (antoineB) [reply grammar.js:965]: done

File: grammar.js (hunk 8 above)
Comment (antoineB) [reply grammar.js:992]: done

File: test/corpus/insert.txt (hunk 9 above)
Comment (dmfay) [reply test/corpus/insert.txt:538]: yeah, double quotes are a MySQL/MariaDBism afaik

**Review by dmfay — APPROVED** [review]
@matthias-Q if you're alright with the current state feel free to land it, or let me know and I can later :rocket:

**Review by matthias-Q — APPROVED** [review]

**Pull request merged**